src/common/base/CollectNSucceeded-inl.h

```diff
@@ -11,59 +11,109 @@ namespace nebula {
 
 // Not thread-safe, all futures need to be on the same executor
 template <class FutureIter, typename ResultEval>
-folly::Future<SucceededResultList<FutureIter>> collectNSucceeded(FutureIter first,
-                                                                 FutureIter last,
-                                                                 size_t n,
-                                                                 ResultEval&& eval) {
+folly::SemiFuture<SucceededResultList<FutureIter>> collectNSucceeded(
+    FutureIter first,
+    FutureIter last,
+    size_t n,
+    ResultEval&& eval) {
   using Result = SucceededResultList<FutureIter>;
   if (n == 0) {
-    return folly::Future<Result>(Result());
+    return folly::SemiFuture<Result>(Result());
   }
 
   struct Context {
-    Context(size_t total, ResultEval&& e) : eval(std::forward<ResultEval>(e)), nTotal(total) {}
+    Context(size_t total, ResultEval&& e)
+        : eval(std::forward<ResultEval>(e))
+        , results(total)
+        , nTotal(total) {}
 
     ResultEval eval;
-    Result results;
-    std::atomic<size_t> numCompleted = {0};
+    std::vector<folly::Optional<folly::Try<FutureReturnType<FutureIter>>>> results;
+    std::atomic<size_t> nCompleted = {0};
     std::atomic<size_t> nSucceeded = {0};
+    std::atomic<bool> fulfilled = false;
     folly::Promise<Result> promise;
-    size_t nTotal;
+    const size_t nTotal;
   };
 
   size_t total = size_t(std::distance(first, last));
   DCHECK_GE(total, 0U);
 
   if (total < n) {
-    return folly::Future<Result>(
+    return folly::SemiFuture<Result>(
         folly::exception_wrapper(std::runtime_error("Not enough futures")));
   }
 
-  auto ctx = std::make_shared<Context>(total, std::forward<ResultEval>(eval));
+  std::vector<folly::futures::detail::DeferredWrapper> executors;
+  folly::futures::detail::stealDeferredExecutors(executors, first, last);
 
   // for each succeeded Future, add to the result list, until
   // we have required number of futures, at which point we fulfil
   // the promise with the result list
+  auto ctx = std::make_shared<Context>(total, std::forward<ResultEval>(eval));
   for (size_t index = 0; first != last; ++first, ++index) {
-    first->setCallback_([n, ctx, index](auto, folly::Try<FutureReturnType<FutureIter>>&& t) {
-      if (!ctx->promise.isFulfilled()) {
+    first->setCallback_([n, ctx, index](auto&&,
+                                        folly::Try<FutureReturnType<FutureIter>>&& t) {
+      VLOG(3) << "Received the callback from the future " << index;
+
+      // relaxed because this guards control but does not guard data
+      auto const c = 1 + ctx->nCompleted.fetch_add(1, std::memory_order_relaxed);
```

Comment:
Maybe you can't use `memory_order_relaxed` here, consider `3 replica, total = 2, n = 1`
1. both thread get `c = 0` (the `1 + ...` is not counted)
2. thread 1 failed because of exception, and return in line 77
3. thread 2 failed, but the we can still pass the `if` statement in line 76, and return in line 77

At this point the promise of `ctx` will never be set.